native/src/main/scala/sqs4s/auth/Credentials.scala

```diff
@@ -0,0 +1,234 @@
+            case (Some(accessKey), Some(secretKey), None) =>
+              BasicCredential(accessKey, secretKey)
+                .pure[F].widen
+
+            case (Some(accessKey), Some(secretKey), Some(token)) =>
+              TemporarySecurityCredential(
+                accessKey,
+                secretKey,
+                token
+              ).pure[F].widen
+
+            case _ =>
+              Sync[F].raiseError[Credential](NoSystemPropertiesFound)
+          }
+        }
+      }
+    }
+
+  /**
+    * Create a resource of temporary credential, credential is automatically
+    * retrieved from EC2 instance metadata: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html#instance-metadata-security-credentials
+    *
+    * @param client        HTTP client
+    * @param ttl           session token's time to live
+    * @param refreshBefore a small duration to refresh the token before it expires
+    * @tparam F an effect which represents the side effects
+    * @return a Resource of TemporarySecurityCredential
+    */
+  def instanceMetadata[F[_]: Concurrent: Clock: Timer](
+    client: Client[F],
+    ttl: FiniteDuration = 6.hours,
+    refreshBefore: FiniteDuration = 5.minutes
+  ): Resource[F, Credentials[F]] = {
+    val dsl = new Http4sClientDsl[F] {}
+    import dsl._
+
+    // Static address to retrieve cred from per API doc: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html#instance-metadata-security-credentials
+    // and AWS SDK: https://github.com/aws/aws-sdk-java/blob/542606ddd5ce6f0c3570ef7488c4f5479c6425bb/aws-java-sdk-core/src/main/java/com/amazonaws/util/EC2MetadataUtils.java#L73
+    val awsLinkLocal = uri"http://169.254.169.254"
+    val tokenEndpoint = awsLinkLocal / "latest" / "api" / "token"
+    val credsEndpoint =
+      awsLinkLocal / "latest" / "meta-data" / "iam" / "security-credentials"
+    val ttlHeader = "X-aws-ec2-metadata-token-ttl-seconds"
+    val tokenHeader = "X-aws-ec2-metadata-token"
+
+    val onError: Response[F] => F[Throwable] = resp => {
+      if (resp.status.responseClass == Status.ServerError) {
+        RetriableServerError.pure[F].widen
+      } else {
+        UnknownAuthError(resp.status).pure[F].widen
+      }
+    }
+
+    val refresh: F[CredentialResponse] =
+      for {
+        token <- client.expectOr[String](PUT(tokenEndpoint))(onError)
+        creds <- client.expectOr[List[CredentialResponse]](
+          GET(
+            credsEndpoint,
+            Header(ttlHeader, ttl.toSeconds.toString),
+            Header(tokenHeader, token)
+          )
+        )(onError)
+        cred <- Sync[F].fromOption(
+          creds.headOption,
+          NoInstanceProfileCredentialFound
+        )
+      } yield cred
+
+    Resource.liftF {
+      refresh.map { init =>
+        Stream
+          .repeatEval(refresh)
+          .metered(ttl - refreshBefore)
+          .holdResource(init).map { sig =>
```

Comment:
@BalmungSan I just realised with the `holdResource` approach, it is not possible to check if the Credential has expired and do a final refresh since we can only do `.get` on `Signal`. Any suggestion?

Reply:
this is what I used to have with `Ref` and `Fiber`:
```
Resource {
        fiberF.map { fiber =>
          val tmp = for {
            resp <- ref.get
            now <- Clock[F].realTime(TimeUnit.MILLISECONDS)
            updated <- if (resp.expiration.toEpochMilli < now) {
              refresh.flatMap(r => ref.updateAndGet(_ => r))
            } else {
              resp.pure[F]
            }
          } yield {
            TemporarySecurityCredential(
              updated.accessKeyId,
              updated.secretAccessKey,
              updated.token
            )
          }
          instance(tmp.widen) -> fiber.cancel
        }
      }
```

Reply:
But do we really need that? I mean the signal should be always updated with some valid credentials.
If the credentials expired before the refresh is because your configuration is wrong, which for me should kill the application. What do you think?

Reply:
I think that is fair. I will stick with `Signal` 👍